File: hbp100/extractors/phones.py
import re
from typing import List

from hbp100.extractors.base import BaseExtractor
from hbp100.schemas.entity import Entity, EntityType


class PhoneExtractor(BaseExtractor):
    def _compile_patterns(self):
        self._patterns = [
            re.compile(r'\b\+\d{1,3}[-.\s]?\(?\d{1,4}\)?[-.\s]?\d{1,4}[-.\s]?\d{1,9}\b'),
            re.compile(r'\b\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b'),
            re.compile(r'\b\d{3}[-.\s]\d{3}[-.\s]\d{4}\b'),
            re.compile(r'\b\d{10}\b'),
        ]

    @property
    def supported_types(self) -> List[EntityType]:
        return [EntityType.PHONE]

    def extract(self, text: str) -> List[Entity]:
        self._validate_text(text)

        entities = []
        detected_digits = set()

        for pattern in self._patterns:
            for match in pattern.finditer(text):
                value = match.group(0)
                cleaned = re.sub(r'[^0-9]', '', value)

                if cleaned in detected_digits:
                    continue

                if 10 <= len(cleaned) <= 15:
                    start, end = match.span()
                    entity = Entity(
                        type=EntityType.PHONE,
                        value=value,
                        start=start,
                        end=end,
                        confidence=0.85,
                    )
                    entities.append(entity)
                    detected_digits.add(cleaned)

        return entities

File: src/extractors/phones.rs
use crate::extractors::base::BaseExtractor;
use crate::interfaces::EntityExtractor;
use crate::schemas::{Entity, EntityType};
use regex::Regex;
use std::collections::HashSet;

pub struct PhoneExtractor {
    patterns: Vec<Regex>,
}

impl PhoneExtractor {
    pub fn new() -> Self {
        let mut extractor = Self {
            patterns: Vec::new(),
        };
        extractor.compile_patterns();
        extractor
    }
}

impl Default for PhoneExtractor {
    fn default() -> Self {
        Self::new()
    }
}

impl EntityExtractor for PhoneExtractor {
    fn extract(&self, text: &str) -> Vec<Entity> {
        if let Err(e) = self.validate_text(text) {
            log::warn!("Validation failed: {}", e);
            return Vec::new();
        }
        
        let mut entities = Vec::new();
        let mut detected_digits = HashSet::new();
        
        for pattern in &self.patterns {
            for caps in pattern.captures_iter(text) {
                if let Some(matched) = caps.get(0) {
                    let value = matched.as_str();
                    let cleaned: String = value.chars().filter(|c| c.is_ascii_digit()).collect();
                    
                    if !detected_digits.contains(&cleaned) && (10..=15).contains(&cleaned.len()) {
                        detected_digits.insert(cleaned);
                        entities.push(Entity::new(
                            EntityType::Phone,
                            value,
                            matched.start(),
                            matched.end(),
                            0.85,
                        ));
                    }
                }
            }
        }
        
        entities
    }
    
    fn supported_types(&self) -> Vec<EntityType> {
        vec![EntityType::Phone]
    }
}

impl BaseExtractor for PhoneExtractor {
    fn compile_patterns(&mut self) {
        self.patterns = vec![
            Regex::new(r"\b\+\d{1,3}[-.\s]?\(?\d{1,4}\)?[-.\s]?\d{1,4}[-.\s]?\d{1,9}\b").unwrap(),
            Regex::new(r"(?<!\w)\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}\b").unwrap(),
            Regex::new(r"\b\d{3}[-.\s]\d{3}[-.\s]\d{4}\b").unwrap(),
            Regex::new(r"\b\d{10}\b").unwrap(),
        ];
    }
}
